src/Gui/CommandDoc.cpp

```diff
@@ -271,44 +272,221 @@ StdCmdExport::StdCmdExport()
-            QString::fromUtf8(QT_TR_NOOP("Please select first the objects you want to export.")));
+            QString::fromUtf8(QT_TR_NOOP("Select the objects to export before choosing Export.")));
         return;
     }
 
-    // fill the list of registered endings
+    // fill the list of registered suffixes
     QStringList filterList;
-    std::map<std::string, std::string> FilterList = App::GetApplication().getExportFilters();
-    std::map<std::string, std::string>::const_iterator jt;
-    for (jt=FilterList.begin();jt != FilterList.end();++jt) {
+    std::map<std::string, std::string> filterMap = App::GetApplication().getExportFilters();
+    for (const auto &filter : filterMap) {
         // ignore the project file format
-        if (jt->first.find("(*.FCStd)") == std::string::npos) {
-            filterList << QString::fromLatin1(jt->first.c_str());
+        if (filter.first.find("(*.FCStd)") == std::string::npos) {
+            filterList << QString::fromStdString(filter.first);
         }
     }
-
     QString formatList = filterList.join(QLatin1String(";;"));
-    Base::Reference<ParameterGrp> hPath = App::GetApplication().GetUserParameter().GetGroup("BaseApp")
-                               ->GetGroup("Preferences")->GetGroup("General");
+    Base::Reference<ParameterGrp> hPath = 
+        App::GetApplication().GetUserParameter().GetGroup("BaseApp")->GetGroup("Preferences")->GetGroup("General");
     QString selectedFilter = QString::fromStdString(hPath->GetASCII("FileExportFilter"));
+    if (!lastExportFilterUsed.isEmpty()) {
+        selectedFilter = lastExportFilterUsed;
+    }
 
+    // Create a default filename for the export
+    // * If this is the first export this session default, generate a new default.
+    // * If this is a repeated export during the same session:
+    //     * If the user accepted the default filename last time, regenerate a new
+    //       default, potentially updating the object label.
+    //     * If not, default to their previously-set export filename.
+    QString defaultFilename = lastExportFullPath;
+
+    bool filenameWasGenerated = false;
+    // We want to generate a new default name in two cases:
+    if (defaultFilename.isEmpty() || lastExportUsedGeneratedFilename) {
+        // First, get the name and path of the current .FCStd file, if there is one:
+        QString docFilename = QString::fromUtf8(
+            App::GetApplication().getActiveDocument()->getFileName());
+
+        // Find the default location for our exported file. Three possibilities:
+        QString defaultExportPath;
+        if (!lastExportFullPath.isEmpty()) {
+            QFileInfo fi(lastExportFullPath);
+            defaultExportPath = fi.path();
+        }
+        else if (!docFilename.isEmpty()) {
+            QFileInfo fi(docFilename);
+            defaultExportPath = fi.path();
+        }
+        else {
+            defaultExportPath = Gui::FileDialog::getWorkingDirectory();

```

Comment:
I'm not sure I understand your comment (which probably means my code is not clear enough): `defaultExportPath` is the location the export window opens up to. That chunk of code is meant to allow a separate "default" directory location for exports than that used for other File dialog operations, on the theory that you will often repeatedly export to the same location, and that may not be the location of your FCStd file or your last non-export file operation. So the first time the export window opens it behaves exactly as FreeCAD does now, but on later opens it can potentially be slightly more intelligent about where it starts you.

Reply:
I think the code is quite clear. I noticed when testing, that if I export before the file as been saved, then the current working directory become the export filename. I believe this behavior comes from this line. This is the behavior that I am unsure of.

Reply:
That doesn't sound like the behavior I expect -- what is supposed to happen is that if there is no filename set, it uses the label instead (which should be something like "Unnamed" or "Unnamed001"). If that's not the case, can you let me know your system details, and also whether you are using the native save dialog, or the QFileDialog?

Reply:
System details in above comment. But interestingly I cannot reproduce it now. So the point is moot I guess. Sorry for the noise.

Reply:
Ah, I got it. It is what you get if you select 'Save a Copy' from the file menu. I must have clicked that by mistake. Still a strange choice, but not your fault I guess :)

Reply:
So infact this behavior upon 'Save a Copy' appears to be a side-effect of your changes here (or some other recent commit). Because I don't see it in the latest AppImage. Let me know if you can reproduce it. And sorry for all the messages :)

Reply:
Thanks - it looks like I need to check to see if a default filename is provided, and don't call `selectFile()` if it's only got a directory.

Reply:
OK, 09476e7 fixes this -- thanks again!